### PR TITLE
476 flows provide api for server side filtering of flows that respects pagination as well

### PR DESCRIPTION
**Pull request by zaychenko-sergei**
## Description

Closes: #476 

Server-side filtering of flows.

## Checklist before requesting a review
- [x] [CHANGELOG.md](./CHANGELOG.md) updated
- [x] API changes are backwards-compatible
- [x] Workspace layout changes include a migration
- [x] Documentation update PR: <link or N/A>
- [x] Dataset pipelines update scheduled if needed

### CODE REVIEW

File: src/adapter/graphql/src/queries/datasets/dataset_flow_runs.rs [hunk 1]
```diff
@@ -62,20 +63,21 @@ impl DatasetFlowRuns {
...
-        let flows_stream = flow_service
-            .list_all_flows_by_dataset(&self.dataset_handle.id)
+        let flows_state_listing = flow_service
+            .list_all_flows_by_dataset(
+                &self.dataset_handle.id,
+                match filters {
+                    Some(filters) => filters.into(),
+                    None => Default::default(),
+                },
+                fs::FlowPaginationOpts { page, per_page },
```
[the remark this answers is not in the document]
Comment (zaychenko-sergei) [reply src/adapter/graphql/src/queries/datasets/dataset_flow_runs.rs:73]: Purely style choice, but let it be

File: src/domain/flow-system/src/entities/flow/flow_id.rs [hunk 2]
```diff
@@ -40,4 +40,7 @@ impl From<FlowID> for u64 {
 pub type FlowIDStream<'a> =
     std::pin::Pin<Box<dyn tokio_stream::Stream<Item = Result<FlowID, InternalError>> + Send + 'a>>;
 
+pub type FlowIDCountFuture<'a> =
+    std::pin::Pin<Box<dyn std::future::Future<Output = usize> + Send + 'a>>;
```
[the remark this answers is not in the document]
Comment (zaychenko-sergei) [reply src/domain/flow-system/src/entities/flow/flow_id.rs:44]: Tried to follow the original style - the data streams methods were sync everywhere and returned pinned boxed streams.

[the remark this answers is not in the document]
Comment (zaychenko-sergei) [reply src/domain/flow-system/src/entities/flow/flow_id.rs:44]: Simplified

File: src/infra/flow-system-inmem/src/services/flow/flow_service_inmem.rs [hunk 3]
```diff
@@ -544,94 +544,104 @@ impl FlowService for FlowServiceInMemory {
...
+            for flow_id in relevant_flow_ids {
                 let flow = Flow::load(flow_id, self.flow_event_store.as_ref()).await.int_err()?;
-
                 yield flow.into();
             }
-        }))
+        });
+
+        let total_count_future =
+            Box::pin(self.flow_event_store.get_count_system_flows(filters_clone));
```
[the remark this answers is not in the document]
Comment (zaychenko-sergei) [reply src/infra/flow-system-inmem/src/services/flow/flow_service_inmem.rs:613]: After getting rid of count future, it looks quite logical:

```
pub struct FlowStateListing<'a> {
    pub matched_stream: FlowStateStream<'a>,
    pub total_count: usize,
}
```

Repository needs separate operations anyway. They might be sequential or might be parallel.
Left sequential calls for now - we first asynchronous wait for count, then stream the results.

From service point of view, it's a holistic structure, higher level of abstraction - a page of listing, where we have data fragment and an indication of total data size.

File: src/infra/flow-system-inmem/src/repos/flow/flow_event_store_inmem.rs [hunk 4]
```diff
@@ -162,131 +281,150 @@ impl FlowEventStore for FlowEventStoreInMem {
...
-                yield flow_id;
+                if matches_filters {
+                    yield flow_id;
+                }
             }
         })
     }
 
+    fn get_count_system_flows(&self, filters: SystemFlowFilters) -> FlowIDCountFuture {
+        Box::pin(async move {
```
[the remark this answers is not in the document]
Comment (zaychenko-sergei) [reply src/infra/flow-system-inmem/src/repos/flow/flow_event_store_inmem.rs:416]: I god rid of async block here, the entire function is async and very simply now.
Note, left it async, as SQL implementation will be async, although, in-memory version on it's own did not have to.

File: src/infra/flow-system-inmem/src/repos/flow/flow_event_store_inmem.rs [hunk 5]
```diff
@@ -162,131 +281,150 @@ impl FlowEventStore for FlowEventStoreInMem {
...
-    fn get_all_flows_by_dataset(&self, dataset_id: &DatasetID) -> FlowIDStream {
-        let dataset_id = dataset_id.clone();
-
+    #[tracing::instrument(level = "debug", skip_all, fields(?filters))]
+    fn get_all_system_flow_ids(
+        &self,
+        filters: SystemFlowFilters,
+        pagination: FlowPaginationOpts,
+    ) -> FlowIDStream {
         // TODO: This should be a buffered stream so we don't lock per record
```
[the remark this answers is not in the document]
Comment (zaychenko-sergei) [reply src/infra/flow-system-inmem/src/repos/flow/flow_event_store_inmem.rs:371]: The influence of `offset`/`limit` is de-facto minimal.
The most important gain is this iteration without per-record locking.

File: src/adapter/graphql/src/queries/datasets/dataset_flow_runs.rs [hunk 6]
```diff
@@ -62,20 +63,25 @@ impl DatasetFlowRuns {
...
+        let matched_flows: Vec<_> = flows_state_listing.matched_stream.try_collect().await?;
+        let total_count = flows_state_listing.total_count;
 
-        let nodes = all_flows
-            .into_iter()
-            .skip(page * per_page)
-            .take(per_page)
-            .map(Flow::new)
-            .collect();
+        let nodes = matched_flows.into_iter().map(Flow::new).collect();
```
Comment (sergiimk) [inline src/adapter/graphql/src/queries/datasets/dataset_flow_runs.rs:84]: if you move `map()` before `matched_stream.try_collect()` you can avoid allocating one extra `Vec`

File: src/adapter/graphql/tests/tests/test_tasks.rs [hunk 7]
```diff
@@ -150,10 +150,11 @@ async fn test_task_list_by_dataset() {
     let mut task_sched_mock = MockTaskScheduler::new();
     task_sched_mock
         .expect_list_tasks_by_dataset()
-        .return_once(move |_| {
-            Ok(Box::pin(futures::stream::iter(
-                [Ok(returned_task)].into_iter(),
-            )))
+        .return_once(move |_, _| {
+            Ok(TaskStateListing {
+                stream: Box::pin(futures::stream::iter([Ok(returned_task)].into_iter())),
```
Comment (sergiimk) [inline src/adapter/graphql/tests/tests/test_tasks.rs:155]: can try [`futures::stream::once`](https://docs.rs/futures/latest/futures/stream/fn.once.html)

File: src/domain/flow-system/src/entities/flow/flow_id.rs [hunk 8]
```diff
@@ -37,7 +35,6 @@ impl From<FlowID> for u64 {
 
 /////////////////////////////////////////////////////////////////////////////////////////
 
-pub type FlowIDStream<'a> =
-    std::pin::Pin<Box<dyn tokio_stream::Stream<Item = Result<FlowID, InternalError>> + Send + 'a>>;
+pub type FlowIDStream<'a> = std::pin::Pin<Box<dyn tokio_stream::Stream<Item = FlowID> + Send + 'a>>;
```
Comment (sergiimk) [inline src/domain/flow-system/src/entities/flow/flow_id.rs:38]: Why did this stream became infallible? Aren't we touching a database for queries like `get_all_flow_ids_by_dataset`?

File: src/infra/flow-system-inmem/src/repos/flow/flow_event_store_inmem.rs [hunk 9]
```diff
@@ -162,160 +281,115 @@ impl FlowEventStore for FlowEventStoreInMem {
...
-                pos -= 1;
-
-                let next = {
-                    let state = self.inner.as_state();
-                    let g = state.lock().unwrap();
-                    g.typed_flows_by_dataset
-                        .get(borrowed_key.as_trait())
-                        .and_then(|flows| flows.get(pos).copied())
-                };
+        let flow_ids_page: Option<Vec<_>> = {
```
Comment (sergiimk) [inline src/infra/flow-system-inmem/src/repos/flow/flow_event_store_inmem.rs:299]: You could do `.collect().unwrap_or_default()` to get rid of `Option<>` which would also eliminate the match below.

**Review by sergiimk — APPROVED** [review]
This is 10x easier to read now :+1: 
Left few minor suggestions.

File: src/adapter/graphql/src/queries/datasets/dataset_flow_runs.rs (hunk 6 above)
Comment (zaychenko-sergei) [reply src/adapter/graphql/src/queries/datasets/dataset_flow_runs.rs:84]: Corrected

File: src/adapter/graphql/tests/tests/test_tasks.rs (hunk 7 above)
Comment (zaychenko-sergei) [reply src/adapter/graphql/tests/tests/test_tasks.rs:155]: Slightly cleaner, but not much shorter:

`stream: Box::pin(futures::stream::once(async { Ok(returned_task) })),`

File: src/infra/flow-system-inmem/src/repos/flow/flow_event_store_inmem.rs (hunk 9 above)
Comment (zaychenko-sergei) [reply src/infra/flow-system-inmem/src/repos/flow/flow_event_store_inmem.rs:299]: Nice

File: src/domain/flow-system/src/entities/flow/flow_id.rs (hunk 8 above)
Comment (zaychenko-sergei) [reply src/domain/flow-system/src/entities/flow/flow_id.rs:38]: My original thought was that only query as a whole coudl fail, but it appears I am wrong after investigating Diesel API.
There are no built-in async streams there , but, let's assume we would use load by iterators.
The return type of such queries is:

```
type RowIter<'conn>: Iterator<Item = QueryResult<U>>...
```

and query result is defined as:

```
enum QueryResult<T> {
    Ok(T),
    Err([Error]),
}
```

So, iterator/stream version should really have result in the items, while queries returning individual items immediately would return `Result`.

**Pull request merged**